src/driver/hermes_cdev.c

```diff
@@ -40,8 +41,121 @@ static struct class *hermes_class;
 DEFINE_IDA(hermes_ida);
 static dev_t hermes_devt;
 
+static long handle_xdma(struct hermes_cmd_req *req,
+			struct hermes_pci_dev *hpdev)
+{
+	struct xdma_cdev *xcdev;
+	uint32_t slot_size;
+	long res;
+	loff_t pos;
+
+	pr_debug("opcode: 0x%x cid: 0x%x slot_type: 0x%x slot_id: 0x%x addr: 0x%llx len: 0x%x\n",
+			req->opcode, req->cid, req->xdma.slot_type,
+			req->xdma.slot_id, req->xdma.addr, req->xdma.len);
+
+	if (req->opcode == HERMES_WR)
+		xcdev = cdev_get_h2c(hpdev);
+	else
+		xcdev = cdev_get_c2h(hpdev);
+
+	if (!xcdev) {
+		pr_err("No DMA channel available");
+		return -HERMES_GENERIC_ERROR;
+	}
+
+	if (req->xdma.slot_type == HERMES_SLOT_PROG) {
+		if (req->xdma.slot_id >= hpdev->cfg.ehpslot) {
+			pr_err("Invalid program slot id: 0x%x\n",
+					req->xdma.slot_id);
+			return -HERMES_STATUS_INV_PROG_SLOT;
+		}
+		pos = hpdev->cfg.ehpsoff;
+		slot_size = hpdev->cfg.ehpssze;
+	} else if (req->xdma.slot_type == HERMES_SLOT_DATA) {
+		if (req->xdma.slot_id >= hpdev->cfg.ehdslot) {
+			pr_err("Invalid data slot id: 0x%x\n",
+					req->xdma.slot_id);
+			return -HERMES_STATUS_INV_DATA_SLOT;
+		}
+		pos = hpdev->cfg.ehdsoff;
+		slot_size = hpdev->cfg.ehdssze;
+	} else {
+		pr_err("Invalid slot type: %#x\n", HERMES_STATUS_INV_SLOT_TYPE);
+		return -HERMES_STATUS_INV_SLOT_TYPE;
+	}
+
+	if (req->xdma.len > slot_size) {
+		pr_err("DMA length greater than slot size: %#x > %#x\n",
+				req->xdma.len, slot_size);
+		return -HERMES_STATUS_NO_SPACE;
+	}
+
+	pos += slot_size * req->xdma.slot_id;
+
+	res = char_sgdma_read_write(xcdev, (char *) req->xdma.addr,
+			req->xdma.len, &pos, req->opcode == HERMES_WR);
+
+	if (res == -EFAULT) {
+		pr_err("Invalid DMA address: %#llx\n", req->xdma.addr);
+		return -HERMES_STATUS_INV_ADDR;
+	} else if (res < 0) {
+		return -HERMES_GENERIC_ERROR;
+	}
+
+	return res;
+}
+
+static int hermes_open(struct inode *inode, struct file *filp)
+{
+	struct hermes_dev *hermes;
+
+	hermes = container_of(inode->i_cdev, struct hermes_dev, cdev);
+	filp->private_data = hermes;
+
+	return 0;
+}
+
+static long hermes_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+	struct hermes_dev *hermes = filp->private_data;
+	struct hermes_pci_dev *hpdev = hermes->hpdev;
+	struct hermes_cmd_req_res *req_res = (struct hermes_cmd_req_res *) arg;
+	struct hermes_cmd_req *req = &req_res->req;
+	struct hermes_cmd_res *res = &req_res->res;
+	long ret = 0;
+
+	switch(req_res->req.opcode) {
+	case HERMES_WR:
+	case HERMES_RD:
+		ret = handle_xdma(req, hpdev);
```

Comment:
Actually, I liked the idea of using reads/writes only where the kernel handles everything.

Something like this:

* On first write, allocate the slots and store them in the file descriptor
* After last write (indicated by something like RWF_SYNC), start the computation
* Use fsync() or RWF_SYNC on last read to finish the job and return the slots.

Any weird reads/writes that are out of order or something should return EINVAL. 

This will be compatible with io_uring (though the only benefit would be to be able to submit multiple jobs to a FD with a single thread).

Reply:
Hmm what if we want to run the same program over multiple data sets or run multiple programs over the same data slot (such as a filter + aggregate)?

I'm thinking we could do something like this:
- We add a new bit to the `Write` command to say if this requires a new slot. The hardware will return the physical slot id to the driver, but doesn't expose it to the user. We can then get rid of the `Request Slot` command.
- For the programmer, the first new program/data it writes will have id 0, the second id 1 and so forth. The driver keeps a map between these ids and the physical slot returned by the device. So the user can say run program X against data Y and it allows the user to release slots.
- For every command written to the device, the driver will keep the response in a FIFO buffer, up to a maximum. If the user doesn't read them, then we start to replace the oldest responses with newer ones.
- When the device is closed, we release all slots still in use.

Reply:
Yes, I forgot the loading for the program. I'd simply suggest that each fd has a single program attached to it.

So before using it you call an IOCTL to attach the program to the fd. It will allocate the appropriate slot and write the program to the device.

If a userspace program wants to interleave multiple accelerator programs then they just need to create multiple fds.

Use reads and writes *only* for data. Don't try to multiplex writes to do both data and attaching the program.

Keep the rule that one fd means one program and one job at a time. Then you don't need any fifo mess and userspace programs that want to do asynchronous things just get more fds.

Reply:
I like the concept of one fd per program, but I don't see how that works for reusing a data slot for different programs. After all, we don't want the user to have to read back + write to new fd between programs.

Reply:
Oh, why would someone want to directly chain programs together though? If you have two programs and one is run on the unmodified output of another, why wouldn't you just compile them into a single program?

Reply:
That's a good point. The only two reasons I can think of are having the programs in binary form only or if compiling them into a single program makes them too large for a slot, but these don't seem good enough. Maybe we can worry about this if/when it becomes an issue. @sbates130272 thoughts?